MediaRecorder should be able to handle audio format change

jyavenard opened the pull request:
#### 6d9edc848d54d1355b4e2b92f28f9692bedd75ad
<pre>
MediaRecorder should be able to handle audio format change
<a href="https://bugs.webkit.org/show_bug.cgi?id=282489">https://bugs.webkit.org/show_bug.cgi?id=282489</a>
<a href="https://rdar.apple.com/139124192">rdar://139124192</a>

Reviewed by Youenn Fablet.

Whenever we encounter an audio sample in a new audio format we will tear down
the current AudioSampleBufferCompressor and create a new one that will perform
the conversion to the original destination format.

Test added.

* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-samplingrate-change-expected.txt: Added.
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-samplingrate-change.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::initialize):
(WebCore::MediaRecorderPrivateEncoder::appendAudioSampleBuffer):
(WebCore::MediaRecorderPrivateEncoder::audioSamplesDescriptionChanged):
(WebCore::MediaRecorderPrivateEncoder::addRingBuffer):
(WebCore::MediaRecorderPrivateEncoder::writeDataToRingBuffer):
(WebCore::MediaRecorderPrivateEncoder::updateCurrentRingBufferIfNeeded):
(WebCore::MediaRecorderPrivateEncoder::audioSamplesAvailable):
(WebCore::MediaRecorderPrivateEncoder::appendVideoFrame):
(WebCore::MediaRecorderPrivateEncoder::partiallyFlushEncodedQueues):
(WebCore::MediaRecorderPrivateEncoder::flushToEndSegment):
(WebCore::MediaRecorderPrivateEncoder::muxNextFrame):
(WebCore::MediaRecorderPrivateEncoder::stopRecording):
(WebCore::MediaRecorderPrivateEncoder::currentTime const):
(WebCore::MediaRecorderPrivateEncoder::currentEndTime const):
(WebCore::MediaRecorderPrivateEncoder::audioCompressor const):
(WebCore::MediaRecorderPrivateEncoder::audioSamplesAvailableStarted): Deleted.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
(WebCore::AudioSampleBufferCompressor::create):
(WebCore::AudioSampleBufferCompressor::AudioSampleBufferCompressor):
(WebCore::AudioSampleBufferCompressor::flushInternal): Have method keeps reference to itself so that we can call flush() and not wait for the output.
(WebCore::AudioSampleBufferCompressor::initAudioConverterForSourceFormatDescription):

Canonical link: <a href="https://commits.webkit.org/286328@main">https://commits.webkit.org/286328@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4e86224c9ef5868b6d85ac340eb27c417a57709

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75678 "Passed style check") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28529 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80164 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26942 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77794 "Passed tests") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2966 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59354 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17531 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78745 "Passed tests") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65004 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39716 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22487 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25271 "Built successfully") | 
| | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22823 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81637 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3017 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1903 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67585 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3168 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64976 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66893 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16668 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10841 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/8984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2974 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5796 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2999 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3934 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3006 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-samplingrate-change.html:
```diff
@@ -0,0 +1,66 @@
+
+        setTimeout(() => {
+            stream.getAudioTracks()[0].applyConstraints({sampleRate:44100});
+        }, 1000);
```
This makes it a really long test. Does it have to go for a whole second before changing?

jyavenard replied:
MediaStream may take a while to start. I found that with a lower value we may not get anything. Most of the MediaRecorder wpt test are over 2.5s each !

jyavenard replied:
I'll lower it to 0.5s

philn commented on LayoutTests/platform/glib/TestExpectations:
```diff
@@ -1162,6 +1162,7 @@ webkit.org/b/213699 imported/w3c/web-platform-tests/mediacapture-record/MediaRec
 webkit.org/b/213699 http/wpt/mediarecorder/set-srcObject-MediaStream-Blob.html [ Pass Failure ]
 webkit.org/b/213699 http/wpt/mediarecorder/MediaRecorder-audio-bitrate.html [ Pass Failure ]
 webkit.org/b/213699 http/wpt/mediarecorder/MediaRecorder-AV-audio-video-dataavailable.html [ Timeout Failure ]
+webkit.org/b/282534 http/wpt/mediarecorder/MediaRecorder-audio-samplingrate-change.html [ Failure ]
```
Is it really failing? Couldn't find any red WPE/GTK EWS test run in the backlog.

jyavenard replied:
Yes. They don’t show as red because I pushed a fix before the tests completed and they shows as cancelled instead. 

look at the one canceled like
https://ews-build.webkit.org/#/builders/34/builds/58736

jyavenard replied:
If they weren’t failing, the expectation would be wrong and it would now fail :)

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
```diff
@@ -39,7 +39,7 @@ namespace WebCore {
 class AudioSampleBufferCompressor : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<AudioSampleBufferCompressor, WTF::DestructionThread::Main> {
 public:
-    static RefPtr<AudioSampleBufferCompressor> create(CMBufferQueueTriggerCallback, void* callbackObject, AudioFormatID = kAudioFormatMPEG4AAC);
+    static RefPtr<AudioSampleBufferCompressor> create(CMBufferQueueTriggerCallback, void* callbackObject, AudioFormatID, std::optional<AudioStreamBasicDescription>);
```
The only call site is passing a AudioStreamBasicDescription, it would simplify things to not use a std::optional.

jyavenard replied:
I plan to use it for audio web codecs

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
```diff
@@ -121,11 +122,7 @@
-        if (!protectedThis)
-            return GenericPromise::createAndReject();
-
+    return invokeAsync(m_serialDispatchQueue, [protectedThis = Ref { *this }, this, isFinished] {
```
Maybe document in the change log that this change is ok. This aligns with VideoEncoder::flush so I think it is fine.

jyavenard replied:
it's required to make sure that all frames will be output when calling flush, even if there's no handler on the promise and we don't have to keep a reference to the AudioSampleBufferCompressor until the flush is done. The MediaRecorderPrivateEncoder makes use of this behaviour

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -277,78 +269,116 @@ void MediaRecorderPrivateEncoder::appendAudioSampleBuffer(const PlatformAudioDat
-        Locker locker { AdoptLock, m_ringBufferLock };
-        if (!m_ringBuffer)
+        Locker locker { m_ringBufferLock };
+        if (m_ringBuffers.isEmpty() || !m_ringBuffers.last())
```
Why these two checks?

jyavenard replied:
In case allocation of the ring buffer failed. We still need to convey the information to handle the error

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:
```diff
@@ -140,15 +141,21 @@ class MediaRecorderPrivateEncoder final : public ThreadSafeRefCountedAndCanMakeT
-    std::optional<CAAudioStreamDescription> m_originalDescription;
     std::optional<std::pair<const MediaTime, GenericPromise::Producer>> m_pendingAudioFramePromise WTF_GUARDED_BY_CAPABILITY(queueSingleton());
+    Lock m_ringBufferLock;
+    Deque<std::unique_ptr<InProcessCARingBuffer>> m_ringBuffers WTF_GUARDED_BY_LOCK(m_ringBufferLock);
```
Why not Deque<UniqueRef<>>?

jyavenard replied:
Allocation of the ring buffer can fail. And we store the null ptr

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -277,78 +269,116 @@ void MediaRecorderPrivateEncoder::appendAudioSampleBuffer(const PlatformAudioDat
+
+    Locker locker { m_ringBufferLock };
+    if (m_ringBuffers.size() > 1)
+        m_ringBuffers.removeFirst();
```
I am not sure to really understand why we use a Deque here. Cannot we just use two std::unique_ptr ring buffers, X and Y?

X would only be used to read and is only set in the queue.
Y would be created where audio samples come from.

On the read side, when we detect that Y is not null, we move Y into X.
On the write side, when Y is not null, we write into Y, otherwise, we write into X.

jyavenard replied:
Theoretically we can have more than one change of format before the task gets dispatched. So two ring buffer may not be enough

jyavenard replied:
It seems that you missed how the Deque is used. 
Say we have 3 format changes occurring.
A | B | C
When A is received, RingBuffer A is created and an `audioSamplesDescriptionChanged` is dispatched to the queue.
Samples with format A, they are added to RingBuffer A, and are all dispatched to `audioSamplesAvailable` to the queue.

Do the same for B and C; none of the tasks dispatched has been run yet. All content has been added to their respective ring buffer.

`audioSamplesDescriptionChanged` for A now runs, it takes the pointer to the first RingBuffer and place it in `m_currentRingBuffer` ; having `m_currentRingBuffer` allows for the reader to pull from the RingBuffer without using the need for the lock.

Now `audioSamplesAvailable` for A is run, fetch the samples from RingBuffer A.

`audioSamplesDescriptionChanged` for B now runs, we can pop the first ring buffer from the deque (which is RingBuffer A), there are no A frames in flight/pending, the next ones will be of B format.

Rinse & repeat.

While it's an unlikely scenario, it is possible. The use of an unlimited sized queue is required.

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
```diff
@@ -164,10 +161,16 @@
-    m_destinationFormat.mFormatID = outputFormatID;
-    m_destinationFormat.mSampleRate = outputFormatID == kAudioFormatOpus ? 48000 : m_sourceFormat.mSampleRate;
-    m_destinationFormat.mChannelsPerFrame = m_sourceFormat.mChannelsPerFrame;
+    if (m_outputDescription)
```
I am not sure we need `m_outputDescription`. Maybe we could just initialise m_destinationFormat to 0 if there is no initial description, and check here for mFormatID or mSampleRate.

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -277,78 +269,118 @@ void MediaRecorderPrivateEncoder::appendAudioSampleBuffer(const PlatformAudioDat
+        if (m_ringBuffers.isEmpty() || !m_ringBuffers.last())
             return;
-        m_ringBuffer->store(downcast<WebAudioBufferList>(data).list(), sampleCount, m_lastEnqueuedAudioSampleCount);
+        m_ringBuffers.last()->store(downcast<WebAudioBufferList>(data).list(), sampleCount, m_currentAudioSampleCount);
```
To make this clearer, I would tend to introduce routines:
```
void addRingBuffer(const AudioStreamDescription&);
void writeDataToRingBuffer(...);
void updateRingBufferIfNeeded();
```
This could also reduce the scope of `m_ringBufferLock ` (which I would rename to `m_ringBuffersLock`) maybe.

jyavenard committed:
MediaRecorder should be able to handle audio format change
https://bugs.webkit.org/show_bug.cgi?id=282489
rdar://139124192

Reviewed by Youenn Fablet.

Whenever we encounter an audio sample in a new audio format we will tear down
the current AudioSampleBufferCompressor and create a new one that will perform
the conversion to the original destination format.

Test added.

* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-samplingrate-change-expected.txt: Added.
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-samplingrate-change.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::initialize):
(WebCore::MediaRecorderPrivateEncoder::appendAudioSampleBuffer):
(WebCore::MediaRecorderPrivateEncoder::audioSamplesDescriptionChanged):
(WebCore::MediaRecorderPrivateEncoder::addRingBuffer):
(WebCore::MediaRecorderPrivateEncoder::writeDataToRingBuffer):
(WebCore::MediaRecorderPrivateEncoder::updateCurrentRingBufferIfNeeded):
(WebCore::MediaRecorderPrivateEncoder::audioSamplesAvailable):
(WebCore::MediaRecorderPrivateEncoder::appendVideoFrame):
(WebCore::MediaRecorderPrivateEncoder::partiallyFlushEncodedQueues):
(WebCore::MediaRecorderPrivateEncoder::flushToEndSegment):
(WebCore::MediaRecorderPrivateEncoder::muxNextFrame):
(WebCore::MediaRecorderPrivateEncoder::stopRecording):
(WebCore::MediaRecorderPrivateEncoder::currentTime const):
(WebCore::MediaRecorderPrivateEncoder::currentEndTime const):
(WebCore::MediaRecorderPrivateEncoder::audioCompressor const):
(WebCore::MediaRecorderPrivateEncoder::audioSamplesAvailableStarted): Deleted.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
(WebCore::AudioSampleBufferCompressor::create):
(WebCore::AudioSampleBufferCompressor::AudioSampleBufferCompressor):
(WebCore::AudioSampleBufferCompressor::flushInternal): Have method keeps reference to itself so that we can call flush() and not wait for the output.
(WebCore::AudioSampleBufferCompressor::initAudioConverterForSourceFormatDescription):

Canonical link: https://commits.webkit.org/286328@main
